src/db_ledger.rs

```diff
@@ -0,0 +1,622 @@
+    }
+}
+
+// The data column family
+#[derive(Default)]
+pub struct DataCf {}
+
+impl DataCf {
+    pub fn key(slot_height: u64, index: u64) -> Vec<u8> {
+        let mut key = vec![0u8; 16];
+        LittleEndian::write_u64(&mut key[0..8], slot_height);
+        LittleEndian::write_u64(&mut key[8..16], index);
+        key
+    }
+
+    pub fn slot_height_from_key(key: &[u8]) -> Result<u64> {
+        let mut rdr = io::Cursor::new(&key[0..8]);
+        let height = rdr.read_u64::<LittleEndian>()?;
+        Ok(height)
+    }
+
+    pub fn index_from_key(key: &[u8]) -> Result<u64> {
+        let mut rdr = io::Cursor::new(&key[8..16]);
+        let index = rdr.read_u64::<LittleEndian>()?;
+        Ok(index)
+    }
+}
+
+impl LedgerColumnFamilyRaw for DataCf {
+    fn handle(&self, db: &DB) -> ColumnFamily {
+        db.cf_handle(DATA_CF).unwrap()
+    }
+}
+
+// The erasure column family
+#[derive(Default)]
+pub struct ErasureCf {}
+
+impl ErasureCf {
+    pub fn key(slot_height: u64, index: u64) -> Vec<u8> {
+        DataCf::key(slot_height, index)
+    }
+}
+
+impl LedgerColumnFamilyRaw for ErasureCf {
+    fn handle(&self, db: &DB) -> ColumnFamily {
+        db.cf_handle(ERASURE_CF).unwrap()
+    }
+}
+
+// ledger window
+pub struct DbLedger {
+    // Underlying database is automatically closed in the Drop implementation of DB
+    pub db: DB,
+    pub meta_cf: MetaCf,
+    pub data_cf: DataCf,
+    pub erasure_cf: ErasureCf,
+}
+
+// TODO: Once we support a window that knows about different leader
+// slots, change functions where this is used to take slot height
+// as a variable argument
+pub const DEFAULT_SLOT_HEIGHT: u64 = 0;
+// Column family for metadata about a leader slot
+pub const META_CF: &str = "meta";
+// Column family for the data in a leader slot
+pub const DATA_CF: &str = "data";
+// Column family for erasure data
+pub const ERASURE_CF: &str = "erasure";
+
+impl DbLedger {
+    // Opens a Ledger in directory, provides "infinite" window of blobs
+    pub fn open(ledger_path: &str) -> Result<Self> {
+        // Use default database options
+        let mut options = Options::default();
+        options.create_if_missing(true);
+        options.create_missing_column_families(true);
+
+        // Column family names
+        let cfs = vec![META_CF, DATA_CF, ERASURE_CF];
+
+        // Open the database
+        let db = DB::open_cf(&options, ledger_path, &cfs)?;
+
+        // Create the metadata column family
+        let meta_cf = MetaCf::default();
+
+        // Create the data column family
+        let data_cf = DataCf::default();
+
+        // Create the erasure column family
+        let erasure_cf = ErasureCf::default();
+
+        Ok(DbLedger {
+            db,
+            meta_cf,
+            data_cf,
+            erasure_cf,
+        })
+    }
+
+    pub fn write_blobs<'a, I>(&mut self, slot: u64, blobs: &'a I) -> Result<()>
+    where
+        &'a I: IntoIterator<Item = &'a &'a Blob>,
```

Comment:
this reads like something I've seen in ledger, block, etc.  I don't fully understand the constraints on I, but I assume it's as flexible as possible (takes vecs, slices, iterators, etc.)

Reply:
The double ref is surprising

Reply:
This was engineered to accept slices xD. So you could pass it references to vectors which coerce down to the appropriate type, or pass a slice. The double ref is due to how slices implement IntoIterator: https://stackoverflow.com/questions/35144386/passing-slice-as-an-intoiterator

Reply:
Amazing, thanks!